### PR TITLE
MAINT: move LSAP rectangular matrix handling into solver code

### PR DESCRIPTION
**Pull request by pmla**
#### What does this implement/fix?

The code for handling tall rectangular cost matrices is split up over `_lsap.py`, `_lsap_module.c`, and the solver code. Here I have moved it all into the solver code.

#### Additional information
I'm doing some maintenance on this module. I am keeping each PR small to make reviewing easier. 

@peterbell10 would you like to take a look again?

### CODE REVIEW

**Review by peterbell10 — COMMENTED** [review]
Can you add some sizes to the benchmarks that hit the transposed case.

File: scipy/optimize/rectangular_lsap/rectangular_lsap.cpp [hunk 1]
```diff
@@ -40,10 +40,21 @@ pseudocode described in pages 1685-1686 of:
...
 #include <vector>
+#include <numeric>
+#include <algorithm>
+#include "rectangular_lsap.h"
+
+
+template <typename T> std::vector<intptr_t> argsort_iter(const std::vector<T> &v)
+{
+    std::vector<intptr_t> index(v.size());
+    iota(index.begin(), index.end(), 0);
```
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/optimize/rectangular_lsap/rectangular_lsap.cpp]: Good suggestion.

File: scipy/optimize/_lsap.py [hunk 2]
```diff
@@ -97,12 +97,4 @@ def linear_sum_assignment(cost_matrix, maximize=False):
...
-    cost_matrix = cost_matrix.astype(np.double)
-
-    # The algorithm expects more columns than rows in the cost matrix.
-    if cost_matrix.shape[1] < cost_matrix.shape[0]:
-        a, b = _lsap_module.calculate_assignment(cost_matrix.T)
-        indices = np.argsort(b)
-        return (b[indices], a[indices])
-    else:
-        return _lsap_module.calculate_assignment(cost_matrix)
+    return _lsap_module.calculate_assignment(cost_matrix.astype(np.double))
```
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/optimize/_lsap.py]: Ah, nice. I didn't know that it also performed casting.

**Review by peterbell10 — APPROVED** [review]
Changes make sense, and benchmark results show improvement even in the non-transposed case. So, this looks very good to me.
```
       before           after         ratio
     [9957042d]       [eb00aaa4]
     <master>         <lsap_transpose>
-     1.05±0.02ms          987±5μs     0.94  optimize_lap.LinearAssignment.time_evaluation((200, 200), 'integer')
-         290±6μs          271±2μs     0.93  optimize_lap.LinearAssignment.time_evaluation((100, 100), 'logarithmic')
-         252±4μs          235±2μs     0.93  optimize_lap.LinearAssignment.time_evaluation((100, 100), 'spatial')
-     4.49±0.02ms      4.18±0.08ms     0.93  optimize_lap.LinearAssignment.time_evaluation((400, 400), 'integer')
-         271±2μs          249±6μs     0.92  optimize_lap.LinearAssignment.time_evaluation((100, 100), 'uniform')
-         233±4μs          212±2μs     0.91  optimize_lap.LinearAssignment.time_evaluation((100, 100), 'integer')
-     1.23±0.08ms      1.05±0.08ms     0.86  optimize_lap.LinearAssignment.time_evaluation((300, 300), 'binary')
-     1.06±0.07ms         897±70μs     0.85  optimize_lap.LinearAssignment.time_evaluation((200, 400), 'binary')
-     1.11±0.01ms         936±70μs     0.84  optimize_lap.LinearAssignment.time_evaluation((400, 200), 'binary')
-      4.42±0.3ms       3.72±0.3ms     0.84  optimize_lap.LinearAssignment.time_evaluation((400, 800), 'binary')
-     2.55±0.03ms       2.14±0.2ms     0.84  optimize_lap.LinearAssignment.time_evaluation((600, 300), 'binary')
-     4.62±0.06ms       3.85±0.3ms     0.83  optimize_lap.LinearAssignment.time_evaluation((800, 400), 'binary')
-        886±60μs         739±70μs     0.83  optimize_lap.LinearAssignment.time_evaluation((200, 400), 'integer')
-         477±3μs          395±1μs     0.83  optimize_lap.LinearAssignment.time_evaluation((200, 200), 'binary')
-         892±4μs         738±60μs     0.83  optimize_lap.LinearAssignment.time_evaluation((400, 200), 'integer')
-     3.68±0.05ms       3.01±0.3ms     0.82  optimize_lap.LinearAssignment.time_evaluation((800, 400), 'integer')
-     1.98±0.03ms       1.61±0.2ms     0.81  optimize_lap.LinearAssignment.time_evaluation((600, 300), 'integer')
-         218±1μs        177±0.5μs     0.81  optimize_lap.LinearAssignment.time_evaluation((200, 100), 'binary')
-       211±0.4μs          167±2μs     0.79  optimize_lap.LinearAssignment.time_evaluation((100, 200), 'binary')
-         148±1μs          115±1μs     0.78  optimize_lap.LinearAssignment.time_evaluation((200, 100), 'integer')
-       154±0.6μs          119±2μs     0.78  optimize_lap.LinearAssignment.time_evaluation((200, 100), 'logarithmic')
-       160±0.3μs        125±0.7μs     0.78  optimize_lap.LinearAssignment.time_evaluation((200, 100), 'spatial')
-         172±1μs          132±1μs     0.77  optimize_lap.LinearAssignment.time_evaluation((100, 200), 'uniform')
-         713±3μs          545±7μs     0.76  optimize_lap.LinearAssignment.time_evaluation((200, 400), 'logarithmic')
-       162±0.7μs          123±3μs     0.76  optimize_lap.LinearAssignment.time_evaluation((100, 200), 'logarithmic')
-      81.0±0.2μs       61.5±0.1μs     0.76  optimize_lap.LinearAssignment.time_evaluation((100, 100), 'binary')
-     1.64±0.01ms      1.24±0.01ms     0.76  optimize_lap.LinearAssignment.time_evaluation((300, 600), 'logarithmic')
-       148±0.7μs        110±0.8μs     0.74  optimize_lap.LinearAssignment.time_evaluation((200, 100), 'uniform')
-       158±0.6μs        117±0.6μs     0.74  optimize_lap.LinearAssignment.time_evaluation((100, 200), 'spatial')
-       143±0.9μs          104±2μs     0.73  optimize_lap.LinearAssignment.time_evaluation((100, 200), 'integer')
-     2.85±0.05ms      2.05±0.01ms     0.72  optimize_lap.LinearAssignment.time_evaluation((400, 800), 'logarithmic')
-      2.11±0.2ms      1.40±0.01ms     0.66  optimize_lap.LinearAssignment.time_evaluation((300, 600), 'spatial')
-        885±10μs          582±3μs     0.66  optimize_lap.LinearAssignment.time_evaluation((400, 200), 'uniform')
-        879±70μs          576±3μs     0.66  optimize_lap.LinearAssignment.time_evaluation((400, 200), 'logarithmic')
-        885±70μs          579±5μs     0.65  optimize_lap.LinearAssignment.time_evaluation((200, 400), 'spatial')
-      1.98±0.2ms      1.28±0.01ms     0.65  optimize_lap.LinearAssignment.time_evaluation((300, 600), 'uniform')
-         854±7μs          551±1μs     0.64  optimize_lap.LinearAssignment.time_evaluation((400, 200), 'spatial')
-      3.87±0.3ms      2.46±0.01ms     0.63  optimize_lap.LinearAssignment.time_evaluation((800, 400), 'logarithmic')
-     2.14±0.02ms      1.34±0.01ms     0.63  optimize_lap.LinearAssignment.time_evaluation((600, 300), 'spatial')
-        834±80μs          521±6μs     0.62  optimize_lap.LinearAssignment.time_evaluation((200, 400), 'uniform')
-     2.03±0.02ms      1.24±0.01ms     0.61  optimize_lap.LinearAssignment.time_evaluation((600, 300), 'uniform')
-      2.00±0.2ms      1.22±0.01ms     0.61  optimize_lap.LinearAssignment.time_evaluation((600, 300), 'logarithmic')
-      4.20±0.2ms      2.50±0.02ms     0.59  optimize_lap.LinearAssignment.time_evaluation((800, 400), 'uniform')
-      4.13±0.2ms      2.41±0.02ms     0.58  optimize_lap.LinearAssignment.time_evaluation((800, 400), 'spatial')
-      3.95±0.5ms      2.17±0.02ms     0.55  optimize_lap.LinearAssignment.time_evaluation((400, 800), 'uniform')
-      3.99±0.5ms      2.17±0.01ms     0.55  optimize_lap.LinearAssignment.time_evaluation((400, 800), 'spatial')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

**Comment by pmla** [comment]
Thanks for another great review @peterbell10

**Pull request merged**